R/crunchy-server.R

```diff
@@ -0,0 +1,101 @@
+#' for specifying an alternate UI for when the user is not authenticated, and
+#' [crunchyUnauthorizedBody()] for giving an alternate UI for users who are
+#' authenticated with Crunch but not authorized to view this app.
+#' @export
+#' @importFrom shiny shinyServer
+crunchyServer <- function (func, authz=getOption("crunchy.authorization")) {
+    shinyServer(function (input, output, session) {
+        public_ui <- getOption("crunchy.body.public", crunchyDefaultPublicUI())
+        output$crunch_body <- renderUI({
+            # First, check whether this user is authenticated.
+            #
+            # Note that this doesn't check the server, just for the existence of
+            # a *.crunch.io token. Generally good enough, and fast. If you
+            # really want to lock it down, specify an authorization function.
+            # E.g. `inherits(shinyUser()(), "UserEntity")` checks that we get a
+            # valid server response that requires authentication
+            if (is.character(input$token) && nchar(input$token) > 0) {
+                # Next, if the user has supplied an authorization requirement,
+                # check that.
+                if (is.null(authz) || is.truthy(authz(input, output, session))) {
+                    # We're good, so call the "normal" server function
+                    func(input, output, session)
+                    # Return the UI body. It is a "lazyUIOutput" that needs to
+                    # be evaluated in the server context.
+                    getOption("crunchy.body")()
+                } else {
+                    # Show the unauthorized page, which defaults to the unauthenticated
+                    getOption("crunchy.body.unauthorized", public_ui)()
+                }
+            } else {
+                # Not authenticated.
+                public_ui()
+            }
+        })
+    })
+}
+
+#' Register authorization logic for your Crunchy app
+#'
+#' Call this to set an expression or server function to evaluate to determine
+#' whether the current user is authorized to access your app. Ideally, this is
+#' cheap to execute because it will be called repeatedly.
+#' @param expr A `function (input, output, session)` or expression to call
+#' inside [crunchyServer()]
+#' @return Invisibly, the server function. If an expression is provided, it will
+#' be wrapped in `function (input, output, session)`. This function is called
+#' for the side effect of setting the authorization function globally.
+#' @export
+#' @examples
+#' \dontrun{
+#' setCrunchyAuthorization(function (input, output, session) {
+#'     # Restrict to users who have crunch.io emails
+#'     endsWith(email(shinyUser()()), "@crunch.io")
+#' })
+#' # This is equivalent; setCrunchyAuthorization will wrap this expression
+#' setCrunchyAuthorization(endsWith(email(shinyUser()()), "@crunch.io"))
+#' }
+setCrunchyAuthorization <- function (expr) {
+    fun <- as.server(expr)
+    options(crunchy.authz=fun)
```

Comment:
Good catch. And that was an unfortunate bug because if no authorization function is found, no check is done. So you might think that you'd restricted access, and instead the app was available to all Crunch users.

A test would've caught that, so I figured out a way to test `crunchyServer` and covered this.